crates/rbuilder-primitives/src/lib.rs

```diff
@@ -877,59 +931,115 @@ impl TransactionSignedEcRecoveredWithBlobs {
+            signer_lookup: None,
+        }
+    }
+}
+
+impl<T: SignerLookup> RawTransactionDecodable<T> {
+    pub fn with_signer_lookup<U: SignerLookup>(self, lookup: U) -> RawTransactionDecodable<U> {
+        RawTransactionDecodable {
+            raw: self.raw,
+            encoding: self.encoding,
+            signer_lookup: Some(lookup),
+        }
+    }
+
+    pub fn decode_enveloped(
+        &self,
+    ) -> Result<TransactionSignedEcRecoveredWithBlobs, TxWithBlobsCreateError> {
+        match self.encoding {
+            TxEncoding::WithBlobData => self.decode_enveloped_with_real_blobs(),
+            TxEncoding::NoBlobData => self.decode_enveloped_with_fake_blobs(),
+        }
+    }
 
     /// Decodes the "raw" format of transaction (e.g. `eth_sendRawTransaction`) with the blob data (network format)
     pub fn decode_enveloped_with_real_blobs(
-        raw_tx: Bytes,
+        &self,
     ) -> Result<TransactionSignedEcRecoveredWithBlobs, TxWithBlobsCreateError> {
-        let raw_tx = &mut raw_tx.as_ref();
+        let raw_tx = &mut self.raw.as_ref();
+
         let pooled_tx = PooledTransactionVariant::decode_2718(raw_tx)
             .map_err(TxWithBlobsCreateError::FailedToDecodeTransaction)?;
-        let signer = pooled_tx
-            .recover_signer()
-            .map_err(|_| TxWithBlobsCreateError::InvalidTransactionSignature)?;
-        match pooled_tx {
-            PooledTransactionVariant::Legacy(_)
-            | PooledTransactionVariant::Eip2930(_)
-            | PooledTransactionVariant::Eip1559(_)
-            | PooledTransactionVariant::Eip7702(_) => {
-                let tx_signed = TransactionSigned::from(pooled_tx);
-                TransactionSignedEcRecoveredWithBlobs::new_no_blobs(tx_signed.with_signer(signer))
-            }
-            PooledTransactionVariant::Eip4844(blob_tx) => {
-                let (blob_tx, signature, hash) = blob_tx.into_parts();
-                let (blob_tx, sidecar) = blob_tx.into_parts();
-                let tx_signed = TransactionSigned::new_unchecked(
-                    Transaction::Eip4844(blob_tx),
-                    signature,
-                    hash,
-                );
-                Ok(TransactionSignedEcRecoveredWithBlobs {
-                    tx: tx_signed.with_signer(signer),
-                    blobs_sidecar: Arc::new(sidecar),
-                    metadata: Metadata::default(),
-                })
+
+        let recover_signer = || {
+            pooled_tx
+                .recover_signer()
+                .map_err(|_| TxWithBlobsCreateError::InvalidTransactionSignature)
+        };
+
+        // TODO: clean this after rust 1.89 and if let chains
+        let signer = if let Some(ref lookup) = self.signer_lookup {
+            if let Some(signer) = lookup.lookup(*pooled_tx.tx_hash()) {
+                signer
+            } else {
+                recover_signer()?
             }
-        }
+        } else {
+            recover_signer()?
+        };
```

Comment:
```suggestion
        let signer = self
            .signer_lookup
            .as_ref()
            .and_then(|lookup| lookup.lookup(*pooled_tx.tx_hash()))
            .or_else(|| pooled_tx.recover_signer().ok())
            .ok_or(TxWithBlobsCreateError::InvalidTransactionSignature)?;
```

Another slightly shorter option

Reply:
[45f9827](https://github.com/flashbots/rbuilder/pull/760/commits/45f98276334057aaa190ffdf839fa8f5717e61a5)